### DISCUSSION TITLE
How to get variable gradient at nodes

### DISCUSSION THREAD

**Discussion by xiaojbing** [idaholab/moose#22098] [Q&A General] [answered]
I use monomial basic function to discrete my variable. I want to get variable value and variable gradient at nodes of current element to compute a coneffient every timestep. 
I try to use getNodalValue() function
```
 _mv_fe_problem(*parameters.getCheckedPointerParam<FEProblemBase *>("_fe_problem_base"))
auto &_variable  = _mv_fe_problem.getStandardVariable(_tid, var_name[i]);
_variable->getNodalValue(*_current_elem->node_ptr(i))
```
to get nodal variable value. 

but I can not find a function to get variable gradient at nodes. How to achieve this?

**Comment by GiudGiud** [comment, marked as answer]
Hello

Are you using the monomial basis function or constant monomials in particular?
Why do you want nodal values? It's often better to work with quadrature point values

Guillaume

**Reply by xiaojbing** [reply]
not using constant monomial. In the iterative calculation process, I do use the orthogonal points, and the purpose of getting the node value is to calculate a coefficient, and the calculation of this coefficient rely on the variable value at the node and the variable gradient at nodes. Finally this coefficient will be coupled into my kernel calculation

**Reply by GiudGiud** [reply]
Seems like all this should happen at quadrature points to me. 
If you want to do it at quadrature points you can use the 
coupledValue and coupledGradient routines from the Coupleable interface. 

Nodes are generally quadrature points for nodal variables so I think this should work. 
Which FE family are you using exactly? what does the variables block look like?

**Reply by xiaojbing** [reply]
just like 
```
[Variables]
    order = SECOND
    family = MONOMIAL
    variables = 'density momx momy momz rhoe'
 []
```

**Reply by GiudGiud** [reply]
umm never used that syntax. I m used to 
```
[Variables]
   [density]
      order = SECOND
      family = MONOMIAL
    []
...
 []
```
